#! /bin/sh /usr/share/dpatch/dpatch-run
## 11-SQUID-2009:1.dpatch by Luigi Gangitano <luigi@debian.org>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Upstream patch fixing Denial of Service in request processing

@DPATCH@
diff -urNad squid3~/src/HttpMsg.cc squid3/src/HttpMsg.cc
--- squid3~/src/HttpMsg.cc	2008-05-03 18:39:08.000000000 +0200
+++ squid3/src/HttpMsg.cc	2009-02-06 20:05:31.000000000 +0100
@@ -463,7 +463,7 @@
 {
 	int i = 0;
 	int retcode = 0;
-	int maj = -1, min = -1;
+	unsigned int maj = 0, min = 0;
 	int last_whitespace = -1, line_end = -1;
 
         debugs(74, 5, "httpParserParseReqLine: parsing " << hmsg->buf);
@@ -568,10 +568,14 @@
 
 			/* next should be 1 or more digits */
 			maj = 0;
-			for (; i < hmsg->req_end && (isdigit(hmsg->buf[i])); i++) {
+			for (; i < hmsg->req_end && (isdigit(hmsg->buf[i])) && maj < 65536; i++) {
 				maj = maj * 10;
 				maj = maj + (hmsg->buf[i]) - '0';
 			}
+                        if (maj >= 65536) {
+                            retcode = -1;
+                            goto finish;
+                        }
 			if (i >= hmsg->req_end) {
 				retcode = 0;
 				goto finish;
@@ -590,10 +594,14 @@
 			/* next should be one or more digits */
 			i++;
 			min = 0;
-			for (; i < hmsg->req_end && (isdigit(hmsg->buf[i])); i++) {
+			for (; i < hmsg->req_end && (isdigit(hmsg->buf[i])) && min < 65536; i++) {
 				min = min * 10;
 				min = min + (hmsg->buf[i]) - '0';
 			}
+                        if (min >= 65536) {
+                            retcode = -1;
+                            goto finish;
+                        }
 
 			/* Find whitespace, end of version */
 			hmsg->v_end = i;
@@ -605,8 +613,6 @@
 	 * Rightio - we have all the schtuff. Return true; we've got enough.
 	 */
 	retcode = 1;
-	assert(maj != -1);
-	assert(min != -1);
 
 finish:
 	hmsg->v_maj = maj;
